install_dev_repos.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+#
+
+"""
+Helper script for installing spyder and external-deps locally in editable mode.
+"""
+
+import argparse
+import os
+import sys
+from logging import Formatter, StreamHandler, getLogger
+from pathlib import Path
+from subprocess import check_output
+
+from importlib_metadata import PackageNotFoundError, distribution
+from packaging.requirements import Requirement
+
+DEVPATH = Path(__file__).resolve().parent
+DEPS_PATH = DEVPATH / 'external-deps'
+BASE_COMMAND = [sys.executable, '-m', 'pip', 'install', '--no-deps']
+
+REPOS = {}
+for p in [DEVPATH] + list(DEPS_PATH.iterdir()):
+    if p.name.startswith('.') or not p.is_dir() and not (
+            (p / 'setup.py').exists() or (p / 'pyproject.toml').exists()):
+        continue
+    try:
+        dist = distribution(p.name)._path
+    except PackageNotFoundError:
+        dist = None
+        editable = None
+    else:
+        editable = (p == dist or p in dist.parents)
+
+    REPOS[p.name] = {'repo': p, 'dist': dist, 'editable': editable}
+
+# ---- Setup logger
+fmt = Formatter('%(asctime)s [%(levelname)s] [%(name)s] -> %(message)s')
+h = StreamHandler()
+h.setFormatter(fmt)
+logger = getLogger('InstallDevRepos')
+logger.addHandler(h)
+logger.setLevel('INFO')
+
+
+def get_python_lsp_version():
+    """Get current version to pass it to setuptools-scm."""
+    req_file = DEVPATH / 'requirements' / 'conda.txt'
+    with open(req_file, 'r', encoding='utf-8') as f:
+        for line in f:
+            if 'python-lsp-server' not in line:
+                continue
+
+            spec = Requirement(line).specifier
+            version = sorted([s.version for s in spec])[0]
+            break
+
+    return version
```

Comment:
```suggestion
            specifiers = Requirement(line).specifier
            break
        else:
            return "0.0.0"
            
    for specifier in specifiers:
        if "=" in specifier.operator:
            return specifier.version
    else:
        return "0.0.0"
```

@mrclary Good catch; the issue you found is due to non-deterministic ordering of the version collection, meaning it was random whether the lower or (non-inclusive) upper bound got selected. In addition, the existing approach is quite hacky, contains several other bugs and won't work correctly for many likely version specifiers. 

Instead, I've come up with and tested the above replacement, which uses the the first version spec that contains "=" (such that it is included in the set of valid versions, and thus a valid version that satisfies the constraints, instead of merely relying on ordering (of course, it still won't work if e.g. `==1.2.*` is specified, but not much we can do there, and we can just make sure to use, e.g. `~=` instead). Additionally, it also returns the default `0.0.0` version in case either the pylsp version is not found in the requirements file, or there is no `=` version with an operator that contains `=` (which should never happen, but it will handle it just in case).